bind_host implementation

toddyocum opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add bind_host setting to set the source IP address for clickhouse client connections.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->
Motivation: For advanced deployment use cases where setting the source IP address used by ClickHouse distributed queries is needed.

Parameters: `<bind_host>X.X.X.X<bind_host> `setting for `<remote_servers>`

Example:
```
<shard>
  <replica>
      <host>172.18.11.224</host>
      <port>9500</port>
      <secure>1</secure>
      <password>foobar</password>
      <bind_host>172.18.55.129</bind_host>
    </replica>
</shard>
```
> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing)

All builds in Builds_1 and Builds_2 stages are always mandatory and will run independently of the checks below:
- [ ] <!---ci_include_stateless--> Only: Stateless tests
- [ ] <!---ci_include_stateful--> Only: Stateful tests
- [ ] <!---ci_include_integration--> Only: Integration tests
- [ ] <!---ci_include_performance--> Only: Performance tests
---
- [ ] <!---ci_exclude_style--> Skip: Style check
- [ ] <!---ci_exclude_fast--> Skip: Fast test
---
- [ ] <!---woolen_wolfdog--> Non-blocking CI mode (Resource-intensive. All test jobs execute in parallel).
- [ ] <!---no_ci_cache--> Disable CI cache

<!--
GitHub Actions can run CI on a PR in one of two ways:
1. Run CI on the branch HEAD.
2. Merge master into the branch HEAD and run CI on the ephemeral merge commit.
Option 2. is safer than 1. but also slower since incoming C++ changes from master typically trash the build artifact cache.
The default in CI is 1. If you like to go for 2. remove the following line:
#no_merge_commit
-->

toddyocum committed:
[WIP] Very draft, partially complete bind_host implementation

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=74741) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=74741) before we can accept your contribution.<br/><hr/>**Todd Yocum** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=74741) it.</sub>

toddyocum commented:
Hi @Michicosun 
 
(Following up on @rschu1ze advice in slack to 'make a little noise' to get some feedback on this PR)

For context, I had briefly discussed this  [FR](https://github.com/ClickHouse/ClickHouse/issues/74587) back in September with @alexey-milovidov  , and at that time he had identified you as maybe a good candidate for working on/with this FR. 

My primary concern about the solution implemented in my PR is I had to modify POCO to expose setting of the bind() for client sockets, so looking for feedback on that first and foremost before I go further with the PR. 

The rest seems fairly straightforward (fingers crossed), since it is mostly about just adding the bind_host setting.

I still need to add the documentation for the new setting, figure out what and how, if any, of tests are reasonable for this feature, and I need to fill out the other code paths, indicated by a TYOCUM TODO comment in the PR.

The PR as-is currently deployed in our labs and has been working as needed for our specific use case, fwiw, and is what we plan  on using in the interim, but really hope to get this PR in by the 25.3 LTS release if possible.

Thanks!

toddyocum committed:
First pass at adding documentation for the new bind_host setting.

rschu1ze commented:
@CheSema Do you like to take a look at this PR? Guess there are not too many people familiar with Poco otherwise 😄

CheSema commented on src/Client/Connection.cpp:
```diff
@@ -152,6 +155,11 @@ void Connection::connect(const ConnectionTimeouts & timeouts)
+                if(!bind_host.empty()) {
+                    Poco::Net::SocketAddress socket_address(bind_host, 0);
+
+                    static_cast<Poco::Net::SecureStreamSocket*>(socket.get())->bind(socket_address, true);
```
Do we need bind source for non secure connection?

toddyocum replied:
Yes, I think so, to be complete. 

The initial implementation was limited to secure connections because that is the only way we use ClickHouse when we deploy it, and I was looking to get feedback on the general approach.

Im working now on updating the PR to include a non-secure connection version, shouldn't be a big change, I think.

toddyocum replied:
The non secure connection version has been implemented now too. 

And the problem breaking the tests has been fixed (was missing a default setting for bind_host, so it was required in config)

What are the next steps now? What issues/concerns remain? 
Thanks!

CheSema replied:
Greate.
I'm OK with the code and with the idea.

The next steps are:

We need to update the docs here. As far as it is new feature, some explanation should be put in the docs.

Also this PR needs test which uses this feature. It could be hard to do actually. 
At first thought I do not know how to set up more than one ip address. 
I'm not sure that I could explain that comprehensive I have the following idea: we could use addresses 127.0.0.{1..127} to reach the same server. And we do that in functional tests. It help us to go to the same server but over the network interfaces. May be we could bind the client on the address 127.0.0.10 and check on the server that only clients from 127.0.0.10 are listened.

toddyocum replied:
Yeah, I'm educating myself on ClickHouse testing in general, and trying to figure out how to test this.

For my own quick dev testing, I've just been setting the bind_host IP to the actual IP, which isn't a great test, but does exercise the features code path and is legal, and then setting the bind_host IP to something non-sensical, like 1.2.3.4, to test the fail condition of the features code path.  

In regards to doc, is there additional documentation I need to do other than the docs I edited in distributed.md?

CheSema committed:
Merge branch 'master' into tyocum/add-bind_host-remote-servers-setting

toddyocum committed:
Style check fixes

toddyocum committed:
Style check fix

maxknv committed:
Merge branch 'master' into tyocum/add-bind_host-remote-servers-setting

toddyocum committed:
Merge branch 'ClickHouse:master' into tyocum/add-bind_host-remote-servers-setting

toddyocum committed:
Merge branch 'master' into tyocum/add-bind_host-remote-servers-setting

toddyocum committed:
Add a default argument of "", the empty string, for the bind_host setting
Add bind() method to StreamSocket
Refactor use of bind_host in Connection::connect() to be used when either secure or not secure

toddyocum committed:
Refactor use of bind_host in Connection::connect() again because of #if USE_SSL

toddyocum committed:
Accidentally missed removal of older code.

toddyocum committed:
Merge branch 'ClickHouse:master' into tyocum/add-bind_host-remote-servers-setting

toddyocum committed:
Remove some draft only comments.

toddyocum committed:
Add some simple tests of bind_host using loopback.

Override is_local setting to make sure the bind_host code paths are hit

More complete test would use tcpdump to verify actual bind_host is being used, although bind_host will fail at call time, so maybe is not necessary to sniff packets if tests are exercising the code paths.

toddyocum committed:
Style fix.

toddyocum committed:
Update 01293_show_clusters_reference to match new output when test_shard_bind_host is included in clusters.xml

toddyocum committed:
Merge branch 'ClickHouse:master' into tyocum/add-bind_host-remote-servers-setting

azat commented:
>Although I'm not sure why the test_shard_bind_host_secure test is failing, it looks like ClickHouse isn't always built with NetSSL support for the CI/CD?

Fast test does not have SSL support, add `Tags: no-fasttest` comment to your test.

toddyocum committed:
Add no-fasttest tag because needs SSL for one of the tests.

toddyocum committed:
Merge branch 'ClickHouse:master' into tyocum/add-bind_host-remote-servers-setting

toddyocum committed:
Merge branch 'ClickHouse:master' into tyocum/add-bind_host-remote-servers-setting

toddyocum committed:
Merge branch 'ClickHouse:master' into tyocum/add-bind_host-remote-servers-setting

toddyocum committed:
Merge branch 'ClickHouse:master' into tyocum/add-bind_host-remote-servers-setting

toddyocum committed:
Merge branch 'ClickHouse:master' into tyocum/add-bind_host-remote-servers-setting

toddyocum committed:
Merge branch 'ClickHouse:master' into tyocum/add-bind_host-remote-servers-setting

toddyocum committed:
Merge branch 'ClickHouse:master' into tyocum/add-bind_host-remote-servers-setting
